components/cli/ballerina/docker_bal_executor.go

```diff
@@ -0,0 +1,147 @@
+ * You may obtain a copy of the License at
+ *
+ * http:www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package ballerina
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/user"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/cellery-io/sdk/components/cli/pkg/constants"
+	"github.com/cellery-io/sdk/components/cli/pkg/util"
+	"github.com/cellery-io/sdk/components/cli/pkg/version"
+)
+
+type DockerBalExecutor struct {
+}
+
+// NewDockerBalExecutor returns a DockerBalExecutor instance.
+func NewDockerBalExecutor() *DockerBalExecutor {
+	balExecutor := &DockerBalExecutor{}
+	return balExecutor
+}
+
+// Build executes ballerina build when ballerina is not installed.
+func (baleExecutor *DockerBalExecutor) Build(fileName string, iName []byte) error {
+	currentDir, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("error in determining working directory, %v", err)
+	}
+	// Retrieve the cellery cli docker instance status.
+	cmdDockerPs := exec.Command("docker", "ps", "--filter",
+		"label=ballerina-runtime="+version.BuildVersion(),
+		"--filter", "label=currentDir="+currentDir, "--filter", "status=running", "--format", "{{.ID}}")
+	containerId, err := cmdDockerPs.Output()
+	if err != nil {
+		return fmt.Errorf("error in retrieving cellery cli docker instance status, %v", err)
+	}
+
+	if string(containerId) == "" {
+		cmdDockerRun := exec.Command("docker", "run", "-d",
+			"-l", "ballerina-runtime="+version.BuildVersion(),
+			"-l", "current.dir="+currentDir,
+			"--mount", "type=bind,source="+currentDir+",target=/home/cellery/src",
+			"--mount", "type=bind,source="+util.UserHomeDir()+string(os.PathSeparator)+".ballerina,target=/home/cellery/.ballerina",
+			"--mount", "type=bind,source="+util.UserHomeDir()+string(os.PathSeparator)+".cellery,target=/home/cellery/.cellery",
+			"wso2cellery/ballerina-runtime:"+version.BuildVersion(), "sleep", "600",
+		)
+		stderrReader, err := cmdDockerRun.StderrPipe()
+		if err != nil {
+			return fmt.Errorf("error while building stderr pipe, %v", err)
+		}
+		stdoutReader, _ := cmdDockerRun.StdoutPipe()
+		if err != nil {
+			return fmt.Errorf("error while building stdout pipe, %v", err)
+		}
+		stderrScanner := bufio.NewScanner(stderrReader)
+		stdoutScanner := bufio.NewScanner(stdoutReader)
+		err = cmdDockerRun.Start()
+		if err != nil {
+			return fmt.Errorf("error while starting docker process, %v", err)
+		}
+		go func() {
+			for {
+				if stderrScanner.Scan() && strings.HasPrefix(stderrScanner.Text(), "Unable to find image") {
+					break
+				}
+			}
+		}()
+		go func() {
+			for {
+				if stdoutScanner.Scan() {
+					containerId = []byte(stdoutScanner.Text())
+					break
```

Comment:
do we need a break statement here?

Reply:
It seems break statement is not necessary, however this was the previous implementation. Removed it.